databases/de-database-schema/src/main/functions/03_app_category_hierarchy.sql

```diff
@@ -68,3 +68,36 @@ TABLE(
     SELECT * FROM subcategories
 $$ LANGUAGE SQL;
 
+--
+-- Yet another version of the same function that allows us to specify the set
+-- of app IDs that the user can access.
+--
+CREATE OR REPLACE FUNCTION app_category_hierarchy(uuid, boolean, anyarray)
+RETURNS
+TABLE(
+    parent_id uuid,
+    id uuid,
+    name varchar(255),
+    description varchar(255),
+    workspace_id uuid,
+    is_public boolean,
+    app_count bigint
+) AS $$
+    WITH RECURSIVE subcategories AS
+    (
+            SELECT acg.parent_category_id AS parent_id, ac.id, ac.name,
+                   ac.description, ac.workspace_id, ac.is_public,
+                   app_count(ac.id, $2, $3) AS app_count
+            FROM app_category_group acg
+            RIGHT JOIN app_category_listing ac ON acg.child_category_id = ac.id
+            WHERE ac.id = $1
+        UNION ALL
+            SELECT acg.parent_category_id AS parent_id, ac.id, ac.name,
+                   ac.description, ac.workspace_id, ac.is_public,
+                   app_count(ac.id, $2, $3) AS app_count
+            FROM subcategories sc, app_category_group acg
+            JOIN app_category_listing ac ON acg.child_category_id = ac.id
+            WHERE acg.parent_category_id = sc.id
+    )
+    SELECT * FROM subcategories
+$$ LANGUAGE SQL;
```

Comment:
The third parameter in this case limits the app counts to apps that the user can view. App categories do not have permissions.